scripts/publish.ts

```diff
@@ -0,0 +1,408 @@
+  }
+  await exec(
+    "yarn",
+    [
+      ...["workspaces", "foreach", "--all"],
+      "--include",
+      `{${[...packageNameToReleaseVersion.keys()].join(",")}}`,
+      ...["--no-private", "--parallel", "--topological-dev"],
+      ...["npm", "publish"],
+      ...["--tag", npmDistributionTag],
+      ...(args["tolerate-republish"] ? ["--tolerate-republish"] : []),
+    ],
+    { dryRun },
+  );
+
+  // Reset the index and working tree.
+  console.log("🧹 Cleaning...");
+  await exec("git", ["reset", "--hard"], { dryRun });
+
+  // Everything went well.
+  if (dryRun) {
+    console.log("This was a dry run.");
+  } else {
+    const successMessage = `Successfully published ${packageNameToReleaseVersion.size} new package${s}.`;
+    console.log(chalk.green(successMessage));
+  }
+
+  process.off("uncaughtException", handleUncaughtException);
+}
+
+function parsePackageReleaseTypeSpecifier(
+  specifier: string,
+): [name: string, releaseType: string] | undefined {
+  const groups = specifier.match(/^(?<name>\S+):(?<releaseType>\S+)$/)?.groups;
+  return groups?.name !== undefined && groups.releaseType !== undefined
+    ? [groups.name, groups.releaseType]
+    : undefined;
+}
+
+function isReleaseType(value: unknown): value is "major" | "minor" | "patch" {
+  return value === "major" || value === "minor" || value === "patch";
+}
+
+function parsePackageVersionSpecifier(
+  specifier: string,
+): [name: string, version: string] | undefined {
+  const groups = specifier.match(/^(?<name>\S+)@(?<version>\S+)$/)?.groups;
+  return groups?.name !== undefined && groups.version !== undefined
+    ? [groups.name, groups.version]
+    : undefined;
+}
+
+function isDefined<T>(value: T): value is NonNullable<T> {
+  return value !== null && value !== undefined;
+}
+
+function findHighestPublishedPrereleaseNumber(
+  packageVersionHistory: string[],
+  parsedPackageVersion: SemVer,
+  prereleaseIdentifier: string,
+): number | undefined {
+  return packageVersionHistory.reduce(
+    (prevHighestPublishedPrereleaseNumber, publishedVersion) => {
+      const parsedPublishedVersion = semverParse(publishedVersion);
+      if (
+        parsedPublishedVersion?.compareMain(parsedPackageVersion) !== 0 ||
+        parsedPublishedVersion.prerelease.slice(0, -1).join(".") !==
+          prereleaseIdentifier
+      ) {
+        // Either the main parts or the prerelease identifiers are different.
+        return prevHighestPublishedPrereleaseNumber;
+      }
+      // `publishedPrereleaseNumber` is the last element of the
+      // `parsedPublishedVersion.prerelease` array.
+      const [publishedPrereleaseNumber] =
+        parsedPublishedVersion.prerelease.slice(-1);
+      return typeof publishedPrereleaseNumber === "number" &&
+        publishedPrereleaseNumber >=
+          (prevHighestPublishedPrereleaseNumber ?? Number.MIN_SAFE_INTEGER)
+        ? publishedPrereleaseNumber
+        : prevHighestPublishedPrereleaseNumber;
```

Comment:
Have you tried using `semver`'s range functions? There may be one that done what this function does, like `semverSort` or `semverMaxSatisfying` (assuming semver's sorting matches what we need).